### PR TITLE
Improve Lale sklearn schema

### PR DESCRIPTION
**Pull request by Ingkarat**
- BaggingClassifier - Add 2 constraints.
- BaggingRegressor - Add 2 constraints.
- ColumnTransformer - Add a sparse constraint.
- ExtraTreesClassifier - Add a constraint.
- ExtraTreesRegressor - Add 2 constraints.
- FeatureAgglomeration - Add a sparse constraint.
- FunctionTransformer - Fix a constraint.
- GaussianNB - Add a sparse constraint.
- KNeighborsClassifier - Remove 2 constraints (false negatives).
- KNeighborsRegressor - Remove 2 constraints (false negatives).
  - For KNN-C and KNN-R, this constraint might be useful but the schema is long and contains many TODOs.

> Metric 'minkowski' not valid for sparse input. Use sorted(sklearn.neighbors.VALID_METRICS_SPARSE['brute']) to get valid options. Metric can also be a callable function.

- LinearSVC - Improve constaints.
- LinearSVR - Modify "loss" schema. Add a constraint.
- LogisticRegression - Add 4 constaints. 
- MinMaxScaler - Add a sparse constraint.
- MissingIndicator - Add a constraint.
- OneHotEncoder - Add "drop" schema, new in 0.21. Add a constraint.
- OrdinalEncoder - Remove "ignore" from "handle_unknown" schema. Add 2 constraints.
- RandomForestClassifier - Add a constraint.
- RandomForestRegressor - Add 2 constraints.
- RidgeClassifier - Add 2 constraints.
- Ridge - Add 2 constraints.
- RobustScaler - Add a constraint.
- SimpleImputer - Add a constraint.
- SVC - Remove a constraint (false negative). Add a constraint.
- SVR - Remove a constraint (false negative). Add a constraint.

### CODE REVIEW

File: lale/lib/sklearn/ordinal_encoder.py [hunk 1]
```diff
@@ -331,5 +331,36 @@ def inverse_transform(self, X):
             set_as_available=True,
         ),
     )
+    OrdinalEncoder = OrdinalEncoder.customize_schema(
+        constraint={
+            "description": "unknown_value should be an integer or np.nan when handle_unknown is 'use_encoded_value'. s",
+            "anyOf": [
```
Comment (hirzel) [inline lale/lib/sklearn/ordinal_encoder.py:337]: The description says that `np.nan` is also okay. So the `"anyOf"` should probably allow for a third alternative:
```python
{
    "type": "object",
    "properties": {"unknown_value": {"enum": [np.nan]}},
},
```

File: lale/lib/sklearn/robust_scaler.py [hunk 2]
```diff
@@ -63,7 +63,14 @@
                     "description": "If False, try to avoid a copy and do inplace scaling instead.",
                 },
             },
-        }
+        },
+        {
+            "description": "From /preprocessing/_data.py:RobustScaler:fit, Exception: raise ValueError(     'Cannot center sparse matrices: use `with_centering=False` instead. See docstring for motivation and alternatives.'     ) ",
```
Comment (hirzel) [inline lale/lib/sklearn/robust_scaler.py:68]: Could you please simplify the description to elide the "From ... raise ..." boiler-plate?

**Review by hirzel — CHANGES_REQUESTED** [review]
Thanks a lot for your PR! It looks great overall. I left a few line-by-line comments.

**Comment by hirzel** [comment]
One more thing. I noticed that there is now a time-out in test.test_optimizers.TestGridSearchCV.test_runtime_limit_hoc. My theory is that this is probably caused by the more elaborate set of constraints for LogisticRegression, which can slow down search space generation. My recommendation is to change `max_opt_time = 2.0` to `max_opt_time = 10.0`. The thinking behind this is that by increasing the time actually spent running trials, the relative proportion of time spent generating the search space goes down.

File: lale/lib/sklearn/ordinal_encoder.py [hunk 3]
```diff
@@ -318,7 +318,7 @@ def inverse_transform(self, X):
 When this parameter is set to `ignore` and an unknown category is encountered during transform,
 the resulting encoding with be set to the value indicated by `encode_unknown_with` (this functionality is added by lale).
 """,
-                values=["error", "ignore", "use_encoded_value"],
```
Comment (kiran-kate) [inline lale/lib/sklearn/ordinal_encoder.py:321]: This might have to be left as is. The value `ignore` was used to apply our own custom fix for an unknown value.

**Comment by hirzel** [comment]
Hi, I have one more request. Is there any chance you could add tests for the newly added constraints? The tests would ensure that they are indeed "true negatives", meaning that they capture problems for which sklearn would also raise an exception. Here is an example for the kind of test I had in mind:
```python
bad_hyperparams = {"bootstrap": False, "oob_score": True}
#without Lale
trainable = sklearn.ensemble.ExtraTreesRegressor(**bad_hyperparams)
with self.assertRaisesRegex(ValueError, "Out of bag estimation only available if bootstrap=True"):
    trained = trainable.fit(X, y)
#with Lale and schema validation
with EnableSchemaValidation():
    with self.assertRaises(jsonschema.ValidationError):
        trainable = lale.lib.sklearn.ExtraTreesRegressor(**bad_hyperparams)
```
Sorry for causing additional work. It's just that we don't users to feel like we are raising spurious exceptions, and these tests would help check for that. A good place for these tests would be [test_type_checking.py](https://github.com/IBM/lale/blob/master/test/test_type_checking.py).

File: lale/lib/sklearn/ordinal_encoder.py (hunk 3 above)
Comment (Ingkarat) [reply lale/lib/sklearn/ordinal_encoder.py:321]: Thank you. I will add it back.

**Comment by Ingkarat** [comment]
Thank you for the comments. I will fix them as well as adding tests.

**Comment by hirzel** [comment]
Excellent, now all tests pass, and all the new constraints are covered by tests that confirm they are "false negatives". Thank you again!!

**Comment by Ingkarat** [comment]
Thank you!
